app/api/items/add_item.go

```diff
@@ -3,145 +3,292 @@ package items
+		service.MustNotBeError(err)
+		return hasAccess
+	})
+}
+
+func constructChildrenValidator(store *database.DataStore, user *database.User) validator.Func {
+	return validator.Func(func(fl validator.FieldLevel) bool {
+		children := fl.Field().Interface().([]itemChild)
+
+		idsMap := make(map[int64]bool, len(children))
+		ids := make([]int64, len(children))
+		for index := range children {
+			idsMap[children[index].ItemID] = true
+			ids[index] = children[index].ItemID
+		}
+		if len(idsMap) != len(children) {
+			return false
+		}
+		hasAccess, err := store.Items().HasManagerAccess(user, ids...)
+		service.MustNotBeError(err)
+		return hasAccess
+	})
+}
+
+func registerValidators(formData *formdata.FormData, store *database.DataStore, user *database.User) {
+	formData.RegisterValidation("parent_item_id", constructParentItemIDValidator(store, user))
+	formData.RegisterTranslation("parent_item_id",
+		"should exist and the user should have manager/owner access to it")
+
+	formData.RegisterValidation("language_id", constructLanguageIDValidator(store))
+	formData.RegisterTranslation("language_id", "no such language")
+
+	formData.RegisterValidation("team_in_group_id", constructTeamInGroupIDValidator(store, user))
+	formData.RegisterTranslation("team_in_group_id", "should exist and be owned by the user")
+
+	formData.RegisterValidation("unlocked_item_ids", constructUnlockedItemIDsValidator(store, user))
+	formData.RegisterTranslation("unlocked_item_ids",
+		"all the IDs should exist and the user should have manager/owner access to them")
+
+	formData.RegisterValidation("children", constructChildrenValidator(store, user))
+	formData.RegisterTranslation("children",
+		"children IDs should be unique and the user should have manager/owner access to them")
+}
+
+func (srv *Service) insertItem(store *database.DataStore, user *database.User, formData *formdata.FormData,
+	newItemRequest *NewItemRequest) (itemID int64, err error) {
+	itemMap := formData.ConstructPartialMapForDB("Item")
+	stringMap := formData.ConstructPartialMapForDB("String")
+
+	err = store.RetryOnDuplicatePrimaryKeyError(func(s *database.DataStore) error {
+		itemID = s.NewID()
+		service.MustNotBeError(s.Items().InsertMap(newItemRequest.itemData(itemID, itemMap)))
+
+		userSelfGroupID, _ := user.SelfGroupID() // the user has been already loaded in addItem()
+		service.MustNotBeError(s.GroupItems().InsertMap(newItemRequest.groupItemData(store.NewID(), user.UserID, userSelfGroupID, itemID)))
+		service.MustNotBeError(s.ItemStrings().InsertMap(newItemRequest.itemStringData(store.NewID(), itemID, stringMap)))
+
+		service.MustNotBeError(
+			s.ItemItems().InsertMap(newItemRequest.itemItemData(store.NewID(), newItemRequest.ParentItemID, itemID, newItemRequest.Order)))
+
+		for _, child := range newItemRequest.Children {
+			service.MustNotBeError(s.ItemItems().InsertMap(newItemRequest.itemItemData(store.NewID(), itemID, child.ItemID, child.Order)))
```

Comment:
isn't there multi-insert in MySQL & gorm?

Reply:
No, there is not. But I rewrote this part to use multi-insert